src/core/qgsmaplayerregistry.cpp

```diff
@@ -151,7 +164,9 @@ void QgsMapLayerRegistry::removeMapLayers( const QList<QgsMapLayer*>& layers )
     QString myId( lyr->id() );
     emit layerWillBeRemoved( myId );
     emit layerWillBeRemoved( lyr );
+    mRegistryLock.lockForWrite();
     mMapLayers.remove( myId );
+    mRegistryLock.unlock();
```

Comment:
I don't think worker threads should modify the registry.
We could insert a Q_ASSERT for this.